File: 程序员面试金典第6版/面试题 16.01. Swap Numbers LCCI.java
// a^b^b = a
// a^b^a = b
// a^0 = a 任何一个数和0异或得到本身
// a^a = 0 任何一个数和本身异或得到0

class Solution {
    public int[] swapNumbers(int[] numbers) {
        numbers[0] = numbers[0]^numbers[1];
        numbers[1] = numbers[0]^numbers[1];
        numbers[0] = numbers[0]^numbers[1];
        return numbers;
    }
}

File: 程序员面试金典第6版/面试题 16.01. Swap Numbers LCCI.cpp
// 加法交换会溢出
// class Solution {
// public:
//     vector<int> swapNumbers(vector<int>& numbers) {
//         numbers[0] = numbers[0] + numbers[1];
//         numbers[1] = numbers[0] - numbers[1];
//         numbers[0] = numbers[0] - numbers[1];
//         return numbers;
//     }
// };
// 减法 100% faster 17.58% memory

class Solution {
public:
    vector<int> swapNumbers(vector<int>& numbers) {
        numbers[0] = numbers[0] - numbers[1];
        numbers[1] = numbers[0] + numbers[1];
        numbers[0] = numbers[1] - numbers[0];
        // numbers[0] = -(numbers[0] - number[1]);
        return numbers;
    }
};